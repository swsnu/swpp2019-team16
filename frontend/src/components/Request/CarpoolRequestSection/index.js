import CarpoolRequestSection from './CarpoolRequestSection';
export default CarpoolRequestSection;
