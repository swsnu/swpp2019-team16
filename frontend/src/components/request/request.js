import React from 'react';
import styled from 'styled-components';
import Button from '../../components/common/Button/Button';
const RequestBlock = styled.div``;

Request.propTypes = {
  //TODO
};

function onCheckboxClickHandler(name, location) {
  return () => {
    var obj = document.getElementsByName(name);
    for (var i = 0; i < obj.length; i++) {
      if (obj[i].value !== location) {
        obj[i].checked = false;
      }
    }
  };
}

function Checkbox(name, value) {
  return (
    <div className="checkbox" key={value}>
      <input
        data-testid={value}
        type="checkbox"
        name={name}
        value={value}
        onClick={onCheckboxClickHandler(name, value)}
      />
      <label>{value}</label>
    </div>
  );
}

function Request({ user, fromList, toList, minimumPassenger, onClickRequest }) {
  const onClickHandler = () => {
    let from = null,
      to = null,
      minPassenger = null;
    var obj = document.getElementsByName('From');
    for (var i = 0; i < obj.length; i++) {
      if (obj[i].checked) {
        from = obj[i].value;
      }
    }

    obj = document.getElementsByName('To');
    for (var j = 0; j < obj.length; j++) {
      if (obj[j].checked) {
        to = obj[j].value;
      }
    }

    obj = document.getElementsByName('minimumPassenger');
    for (var k = 0; k < obj.length; k++) {
      if (obj[k].checked) {
        minPassenger = obj[k].value;
      }
    }

    if (from !== null && to !== null && minPassenger !== null) {
      onClickRequest({
        userId: user.id,
        from: from,
        to: to,
        minimumPassenger: minPassenger,
      });
    } else {
      alert('Please Check!');
    }
  };

  const fromData = fromList.map(function(location) {
    return Checkbox('From', location);
  });

  const toData = toList.map(function(location) {
    return Checkbox('To', location);
  });

  const minPassenger = minimumPassenger.map(function(num) {
    return Checkbox('minimumPassenger', num);
  });

  if (!user) {
    return null;
  }

  return (
    <RequestBlock>
      <div data-testid="from">
        <h2>From</h2>
        {fromData}
      </div>

      <div data-testid="to">
        <h2>To</h2>
        {toData}
      </div>

      <div data-testid="minimumPassenger">
        <h2>Minimum Passenger</h2>
        {minPassenger}
      </div>

      <Button
        id="request-submit-button"
        onClick={onClickHandler}
        children="Carpool Request"
      />
    </RequestBlock>
  );
}

export default Request;
