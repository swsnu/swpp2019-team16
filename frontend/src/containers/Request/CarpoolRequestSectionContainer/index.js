import RequestContainer from './CarpoolRequestSectionContainer';
export default RequestContainer;
