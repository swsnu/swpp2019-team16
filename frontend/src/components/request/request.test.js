import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Request from './request';

describe('<Request />', () => {
  it('SHOULD match with snapshot WHEN user is null', async () => {
    const USER = null;
    const from_list = [];
    const to_list = [];
    const minPassenger = [];

    const { container } = render(
      <Request
        user={USER}
        fromList={from_list}
        toList={to_list}
        minimumPassenger={minPassenger}
        onClickRequest={() => {}}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('SHOULD match with snapshot WHEN all data in', async () => {
    const USER = { id: 1, name: 'zeroFruit' };
    const from_list = ['a', 'b', 'c'];
    const to_list = ['x', 'y', 'c'];
    const minPassenger = ['2', '3', '4'];
    window.alert = () => {
      return true;
    };

    const { container } = render(
      <Request
        user={USER}
        fromList={from_list}
        toList={to_list}
        minimumPassenger={minPassenger}
        onClickRequest={() => {}}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('Button Click without checking checkbox', async () => {
    const USER = { id: 1, name: 'zeroFruit' };
    const from_list = ['a', 'b', 'c'];
    const to_list = ['x', 'y', 'c'];
    const minPassenger = ['2', '3', '4'];
    const mockFn = jest.fn();

    const utils = render(
      <Request
        user={USER}
        fromList={from_list}
        toList={to_list}
        minimumPassenger={minPassenger}
        onClickRequest={mockFn}
      />,
    );

    const button = utils.getByText('Carpool Request');
    fireEvent.click(button);
    expect(mockFn).toHaveBeenCalledTimes(0);
  });

  it('Check the checkbox', async () => {
    const USER = { id: 1, name: 'zeroFruit' };
    const from_list = ['a', 'b', 'c'];
    const to_list = ['x', 'y', 'c'];
    const minPassenger = ['2', '3', '4'];
    const mockFn = jest.fn();

    const utils = render(
      <Request
        user={USER}
        fromList={from_list}
        toList={to_list}
        minimumPassenger={minPassenger}
        onClickRequest={mockFn}
      />,
    );

    const checkbox_from = utils.getByTestId('a');
    fireEvent.click(checkbox_from);
    const checkbox_to = utils.getByTestId('x');
    fireEvent.click(checkbox_to);
    const checkbox_option = utils.getByTestId('2');
    fireEvent.click(checkbox_option);

    const button = utils.getByText('Carpool Request');
    fireEvent.click(button);
    expect(mockFn).toHaveBeenCalledTimes(1);
  });
});
