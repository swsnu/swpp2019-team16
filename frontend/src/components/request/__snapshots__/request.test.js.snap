// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Request /> SHOULD match with snapshot WHEN all data in 1`] = `
<div>
  <div
    class="sc-bdVaJa bDWFJH"
  >
    <div
      data-testid="from"
    >
      <h2>
        From
      </h2>
      <div
        class="checkbox"
      >
        <input
          data-testid="a"
          name="From"
          type="checkbox"
          value="a"
        />
        <label>
          a
        </label>
      </div>
      <div
        class="checkbox"
      >
        <input
          data-testid="b"
          name="From"
          type="checkbox"
          value="b"
        />
        <label>
          b
        </label>
      </div>
      <div
        class="checkbox"
      >
        <input
          data-testid="c"
          name="From"
          type="checkbox"
          value="c"
        />
        <label>
          c
        </label>
      </div>
    </div>
    <div
      data-testid="to"
    >
      <h2>
        To
      </h2>
      <div
        class="checkbox"
      >
        <input
          data-testid="x"
          name="To"
          type="checkbox"
          value="x"
        />
        <label>
          x
        </label>
      </div>
      <div
        class="checkbox"
      >
        <input
          data-testid="y"
          name="To"
          type="checkbox"
          value="y"
        />
        <label>
          y
        </label>
      </div>
      <div
        class="checkbox"
      >
        <input
          data-testid="c"
          name="To"
          type="checkbox"
          value="c"
        />
        <label>
          c
        </label>
      </div>
    </div>
    <div
      data-testid="minimumPassenger"
    >
      <h2>
        Minimum Passenger
      </h2>
      <div
        class="checkbox"
      >
        <input
          data-testid="2"
          name="minimumPassenger"
          type="checkbox"
          value="2"
        />
        <label>
          2
        </label>
      </div>
      <div
        class="checkbox"
      >
        <input
          data-testid="3"
          name="minimumPassenger"
          type="checkbox"
          value="3"
        />
        <label>
          3
        </label>
      </div>
      <div
        class="checkbox"
      >
        <input
          data-testid="4"
          name="minimumPassenger"
          type="checkbox"
          value="4"
        />
        <label>
          4
        </label>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-text"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Carpool Request
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`<Request /> SHOULD match with snapshot WHEN user is null 1`] = `<div />`;
